### PR TITLE
Add filtering by entity type in `getEntities`

### LINKED ISSUE #780: Allow entries API to filter by entity type

### PR DESCRIPTION
**Pull request by rymurr**
This requires the following:
1) add query param to getEntities for which types we want (empty list means all)
2) version store now returns a tuple of Key, entity type
3) in memory and jgit store and populate types on return. Tiered store does not and requires extra lookups

For follow-up PR:
* possibly push filtering into version store. Can potentially be more efficient depending on store impl
* add an entity-type-id to the key list in tiered version store to make it more efficient (#836)

Closes #780

---
This change is [Reviewable](https://reviewable.io/reviews/projectnessie/nessie/837)

### CODE REVIEW

**Comment by jacques-n** [comment]
I'm uncomfortable introducing this api before we've implemented a good way to support it in tiered version store. 

By solving the storage side before exposing at the api level we know that we aren't exposing things in an awkward or inefficient way at the higher levels.

**Comment by rymurr** [comment]
> I'm uncomfortable introducing this api before we've implemented a good way to support it in tiered version store.

Happy either way, I think the discussion on #834 makes it easier to do this.

File: model/src/main/java/org/projectnessie/api/TreeApi.java [hunk 1]
```diff
@@ -126,8 +126,11 @@ public EntriesResponse getEntries(
       @Pattern(regexp = Validation.REF_NAME_OR_HASH_REGEX, message = Validation.REF_NAME_OR_HASH_MESSAGE)
       @Parameter(description = "name of ref to fetch from")
       @PathParam("ref")
-          String refName)
-          throws NessieNotFoundException;
+          String refName,
+      @Parameter(description = "list of value types to return. Return all if empty")
```
Comment (laurentgo) [inline model/src/main/java/org/projectnessie/api/TreeApi.java:130]: It feels like a more generic `filter` query parameter might allow for extensibility

File: servers/services/pom.xml [hunk 2]
```diff
@@ -60,5 +60,11 @@
       <groupId>jakarta.enterprise</groupId>
       <artifactId>jakarta.enterprise.cdi-api</artifactId>
     </dependency>
+      <dependency>
```
Comment (laurentgo) [inline servers/services/pom.xml:63]: I'm not sure why this dependency is being added as I don't see a code change using it, `datanucleus` which is something for DB persistence should not be required

Comment (rymurr) [reply servers/services/pom.xml:63]: IDE just being dumb and helpfully adding deps for me. Will remove it.

File: servers/services/src/main/java/org/projectnessie/services/rest/TreeResource.java [hunk 3]
```diff
@@ -188,18 +189,44 @@ public void mergeRefIntoBranch(String branchName, String hash, Merge merge) thro
...
     } catch (ReferenceNotFoundException e) {
       throw new NessieNotFoundException(String.format("Unable to find the reference [%s].", refName), e);
     }
   }
 
+
+  private static Type toType(String valueType) {
+    switch (valueType) {
+      case "IcebergTable":
+      case "ImmutableIcebergTable":
```
Comment (laurentgo) [inline servers/services/src/main/java/org/projectnessie/services/rest/TreeResource.java:211]: That seems wrong that 2 different value types point to the same Type object

File: model/src/main/java/org/projectnessie/api/TreeApi.java (hunk 1 above)
Comment (rymurr) [reply model/src/main/java/org/projectnessie/api/TreeApi.java:130]: How do you mean? Just rename the parameter or some other object type?

File: versioned/tiered/tiered-impl/src/main/java/org/projectnessie/versioned/impl/TieredVersionStore.java [hunk 4]
```diff
@@ -435,7 +436,7 @@ public void assign(NamedRef namedRef, Optional<Hash> currentTarget, Hash targetH
   }
 
   @Override
-  public Stream<Key> getKeys(Ref ref) throws ReferenceNotFoundException {
+  public Stream<WithEntityType<Key>> getKeys(Ref ref) throws ReferenceNotFoundException {
```
Comment (laurentgo) [inline versioned/tiered/tiered-impl/src/main/java/org/projectnessie/versioned/impl/TieredVersionStore.java:439]: Wouldn't this change back once the filter is pushed down?

File: servers/services/src/main/java/org/projectnessie/services/rest/TreeResource.java (hunk 3 above)
Comment (rymurr) [reply servers/services/src/main/java/org/projectnessie/services/rest/TreeResource.java:211]: agreed, I think based on conversations on #834 we will likely change this

File: versioned/tiered/tiered-impl/src/main/java/org/projectnessie/versioned/impl/TieredVersionStore.java (hunk 4 above)
Comment (rymurr) [reply versioned/tiered/tiered-impl/src/main/java/org/projectnessie/versioned/impl/TieredVersionStore.java:439]: no, as the type is still pushed to the client

File: model/src/main/java/org/projectnessie/api/TreeApi.java (hunk 1 above)
Comment (laurentgo) [reply model/src/main/java/org/projectnessie/api/TreeApi.java:130]: some object type

Comment (rymurr) [reply model/src/main/java/org/projectnessie/api/TreeApi.java:130]: Sorry @laurentgo, its still not clear to me what you are asking for.

**Comment by rymurr** [comment]
This builds on #866 to add filtering to the REST API. Should be ready to go once that is merged

**Review by snazy — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/projectnessie/nessie/837)* status: 0 of 13 files reviewed, 8 unresolved discussions (waiting on @laurentgo and @rymurr)

___
*[python/pynessie/_endpoints.py, line 149 at r2](https://reviewable.io/reviews/projectnessie/nessie/837#-MX2ANgABHVQ3L9SlatS:-MX2ANgABHVQ3L9SlatT:b-v2s8pj) ([raw file](https://github.com/projectnessie/nessie/blob/2c5af2d78304772c394a78adb41f82bcec99653e/python/pynessie/_endpoints.py#L149)):*
> ```Python
>     max_result_hint: Optional[int] = None,
>     page_token: Optional[str] = None,
>     entity_types: Optional[list] = None,
> ```

Mind renaming it to `types` to make the name consistent w/ `TreeApi`?
___
*[python/pynessie/_endpoints.py, line 162 at r2](https://reviewable.io/reviews/projectnessie/nessie/837#-MX2A7A4Ex-b9J0EPnln:-MX2A7A4Ex-b9J0EPnlo:b-lcgdn7) ([raw file](https://github.com/projectnessie/nessie/blob/2c5af2d78304772c394a78adb41f82bcec99653e/python/pynessie/_endpoints.py#L162)):*
> ```Python
>     :return: json list of Nessie table names
>     """
>     if not entity_types:
> ```

I think, we can entirely omit the `types` parameter, when it's not present/empty.
___
*[servers/quarkus-server/src/test/java/org/projectnessie/server/TestRest.java, line 281 at r2](https://reviewable.io/reviews/projectnessie/nessie/837#-MX2BsRgDX8yhtwdHrAx:-MX2BsRgDX8yhtwdHrAy:b4irxib) ([raw file](https://github.com/projectnessie/nessie/blob/2c5af2d78304772c394a78adb41f82bcec99653e/servers/quarkus-server/src/test/java/org/projectnessie/server/TestRest.java#L281)):*
> ```Java
> 
>     entries = tree.getEntries(branch, null, null, ImmutableList.of(Contents.Type.VIEW.name())).getEntries();
>     assertEquals(1, entries.size());
> ```

Nit: These two asserts can be simplified to `assertEquals(Collections.singletonList(expected.get(1)), entries)`
___
*[servers/services/src/main/java/org/projectnessie/services/rest/TreeResource.java, line 223 at r2](https://reviewable.io/reviews/projectnessie/nessie/837#-MX2Aq9v3sLJZ_uQ9oMv:-MX2Aq9v3sLJZ_uQ9oMw:b-vot2ol) ([raw file](https://github.com/projectnessie/nessie/blob/2c5af2d78304772c394a78adb41f82bcec99653e/servers/services/src/main/java/org/projectnessie/services/rest/TreeResource.java#L223)):*
> ```Java
>     //  a more sophisticated approach, potentially with support from the (tiered-)version-store.
>     try {
>       // note currently we are filtering types at the REST level. This could in theory be pushed down to the store though
> ```

Can you move that to the "big TODO" above, so we don't forget about it?

**Review by snazy — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/projectnessie/nessie/837)* status: 0 of 13 files reviewed, 9 unresolved discussions (waiting on @laurentgo and @rymurr)

*[a discussion](https://reviewable.io/reviews/projectnessie/nessie/837#-MX2CcM7BIKXU7kp3Lbz:-MX2CcM7BIKXU7kp3Lc-:b-og9arw) (no related file):*
LGTM, few minor comments

---

**Review by rymurr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/projectnessie/nessie/837)* status: 0 of 13 files reviewed, 8 unresolved discussions (waiting on @laurentgo, @rymurr, and @snazy)

___
*[python/pynessie/_endpoints.py, line 149 at r2](https://reviewable.io/reviews/projectnessie/nessie/837#-MX2ANgABHVQ3L9SlatS:-MX2Dp6y7fe9UwoRSoHl:b-896fix) ([raw file](https://github.com/projectnessie/nessie/blob/2c5af2d78304772c394a78adb41f82bcec99653e/python/pynessie/_endpoints.py#L149)):*
<details><summary><i>Previously, snazy (Robert Stupp) wrote…</i></summary><blockquote>

Mind renaming it to `types` to make the name consistent w/ `TreeApi`?
</blockquote></details>

Done.
___
*[python/pynessie/_endpoints.py, line 162 at r2](https://reviewable.io/reviews/projectnessie/nessie/837#-MX2A7A4Ex-b9J0EPnln:-MX2DpuoCYTHHBgIGxH_:b-896fix) ([raw file](https://github.com/projectnessie/nessie/blob/2c5af2d78304772c394a78adb41f82bcec99653e/python/pynessie/_endpoints.py#L162)):*
<details><summary><i>Previously, snazy (Robert Stupp) wrote…</i></summary><blockquote>

I think, we can entirely omit the `types` parameter, when it's not present/empty.
</blockquote></details>

Done.
___
*[servers/services/src/main/java/org/projectnessie/services/rest/TreeResource.java, line 223 at r2](https://reviewable.io/reviews/projectnessie/nessie/837#-MX2Aq9v3sLJZ_uQ9oMv:-MX2Dr2p7fn_7-AIebPx:b-896fix) ([raw file](https://github.com/projectnessie/nessie/blob/2c5af2d78304772c394a78adb41f82bcec99653e/servers/services/src/main/java/org/projectnessie/services/rest/TreeResource.java#L223)):*
<details><summary><i>Previously, snazy (Robert Stupp) wrote…</i></summary><blockquote>

Can you move that to the "big TODO" above, so we don't forget about it?
</blockquote></details>

Done.

**Review by snazy — APPROVED** [review]

**Pull request merged**